### PR TITLE
azurerm_key_vault_{key|secret|certificate} - Cache the key vault id to url mapping during C/U and avoid calling resource list API during C/U

### PR DESCRIPTION
**Pull request by magodo**
A follow up of #17407, which is related to #11059.

With #17407, the resource list API is avoided in following cases:

- Create a key together with a key vault (managed resource or data source)
- Read a key item together with a key vault (managed resource or data source)

But doesn't cover following cases (as the cache is only populated when a key vault is created/read):

- Create a key separately (since create will call read)
- Read a key separately (this only applies when the depending key vault is hardcoded)
- Update a key separately
- Delete a key (no matter separately or together with the key vault)

This PR made following changes:

1. In `BaseUriForKeyVault` to lookup the key vault base uri by its name, if no hit, populate the record into the cache. This is called in create, which makes the read called by create can read the record back from cache.
2. In the update ~and delete~ function of the key, instead of calling `KeyVaultIDFromBaseUrl()` to get the key vault id, just directly read the `key_vault_id` from state file, as state file is accessble during update and delete, also `key_vault_id` is a force new and required property, which means it won't change between last write to state and this point.

So hopefully, the only time that the resource list api will be called is when the key vault nested item resource are depending on a hardcoded key vault id, or it is deleted.

[Updated] The read (refresh/plan/apply's refresh step) is still likely to suffer this issue. See https://github.com/hashicorp/terraform/issues/31437#issuecomment-1184522278.

I've locally tested for the key, secret and certificate's update acctests, and inspected the API sequence, found only after the deletion, there is the resource list API call, which presumably is from the `Exists()` check in acctest.

### CODE REVIEW

File: internal/services/keyvault/key_vault_certificate_resource.go [hunk 1]
```diff
@@ -681,15 +688,7 @@ func resourceKeyVaultCertificateDelete(d *pluginsdk.ResourceData, meta interface
...
-	keyVaultIdRaw, err := keyVaultsClient.KeyVaultIDFromBaseUrl(ctx, resourcesClient, id.KeyVaultBaseUrl)
-	if err != nil {
-		return fmt.Errorf("retrieving the Resource ID the Key Vault at URL %q: %s", id.KeyVaultBaseUrl, err)
-	}
-	if keyVaultIdRaw == nil {
-		return fmt.Errorf("Unable to determine the Resource ID for the Key Vault at URL %q", id.KeyVaultBaseUrl)
-	}
-
-	keyVaultId, err := parse.VaultID(*keyVaultIdRaw)
+	keyVaultId, err := parse.VaultID(d.Get("key_vault_id").(string))
```
Comment (tombuildsstuff) [inline internal/services/keyvault/key_vault_certificate_resource.go:691]: unfortunately we can't use schema fields in the delete function - this isn't always available in the Plugin SDK and isn't available in Framework so we have to look this up from the API instead

File: internal/services/keyvault/key_vault_secret_resource.go [hunk 2]
```diff
@@ -386,14 +379,7 @@ func resourceKeyVaultSecretDelete(d *pluginsdk.ResourceData, meta interface{}) e
...
 
-	keyVaultIdRaw, err := keyVaultsClient.KeyVaultIDFromBaseUrl(ctx, resourcesClient, id.KeyVaultBaseUrl)
-	if err != nil {
-		return fmt.Errorf("retrieving the Resource ID the Key Vault at URL %q: %s", id.KeyVaultBaseUrl, err)
-	}
-	if keyVaultIdRaw == nil {
-		return fmt.Errorf("Unable to determine the Resource ID for the Key Vault at URL %q", id.KeyVaultBaseUrl)
-	}
-	keyVaultId, err := parse.VaultID(*keyVaultIdRaw)
+	keyVaultId, err := parse.VaultID(d.Get("key_vault_id").(string))
```
Comment (tombuildsstuff) [inline internal/services/keyvault/key_vault_secret_resource.go:382]: (same here)

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hey @magodo 

Thanks for this PR - I've taken a look through and left a few comments inline, but unfortunately we can't use `d.Get` inside of the Delete function - depending on how resources are deleted (`terraform destroy` verses removing them from the config and running `terraform apply`, for example) the config isn't necessarily available during the Delete function. In addition unfortunately terraform-plugin-framework doesn't expose the config during the deletion, so this isn't forwards compatible there - as such we're going to need to continue looking this up during the deletion.

Thanks!

**Comment by magodo** [comment]
@tombuildsstuff My intent here is to read from the state ([prior state](https://github.com/hashicorp/terraform-plugin-go/blob/df65cd885837c0a41f571da95b81a705acc130eb/tfprotov5/internal/tfplugin5/tfplugin5.proto#L291) actually), which should be always available for update and delete. From my understanding, the `d.Get()` will [read field from a couple of sources](https://github.com/hashicorp/terraform-provider-azurerm/blob/d2111e647efca5d9629a933d1bc89b8a93d714a1/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema/resource_data.go#L535-L543). So for this case, it will failed to read from reading from `set` and `diff`, also if the `config` is not available, then it will finally read from `state`, as expected. This is just my assumption, I'll test for this scenario to confirm.

**Comment by magodo** [comment]
@tombuildsstuff I've just verified that using `d.Get()` also works when *removing them from the config and running terraform apply*.

**Comment by tombuildsstuff** [comment]
@magodo 

> I've just verified that using d.Get() also works when removing them from the config and running terraform apply.

In which version of Terraform? At any rate, since that information isn't available in Terraform Plugin Framework we'd end up having to revert this change in time - so unfortunately we can't pull those from the config/state here / would still have to look these up unfortunately.

**Comment by magodo** [comment]
@tombuildsstuff I'm using v1.2.4. For the framework, both [Update](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework@v0.9.0/tfsdk#UpdateResourceRequest) and [Delete](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework@v0.9.0/tfsdk#DeleteResourceRequest) request has access to the prior state.

File: internal/services/keyvault/client/helpers.go [hunk 3]
```diff
@@ -45,6 +45,10 @@ func (c *Client) BaseUriForKeyVault(ctx context.Context, keyVaultId parse.VaultI
 	lock[cacheKey].Lock()
 	defer lock[cacheKey].Unlock()
 
+	if v, ok := keyVaultsCache[cacheKey]; ok {
+		return &v.dataPlaneBaseUri, nil
+	}
```
[the remark this answers is not in the document]
Comment (magodo) [reply internal/services/keyvault/client/helpers.go:50]: Moving it above the lock might introduce concurrent read and write on the `keyVaultsCache` map?

**Comment by tombuildsstuff** [comment]
@magodo apologies for the delay with this one, any chance you could rebase this and we should be able to get this one in?

**Comment by magodo** [comment]
@tombuildsstuff Thank you for revisiting this! I've merged against main branch and verified that running acctest `TestAccKeyVaultSecret_update` will only call one time the list api (due to exists check at the very end).

**Pull request merged**